waltz-test/src/main/python/waltz_ducktape/tests/validation/scalability_test.py

```diff
@@ -97,3 +109,66 @@ def scale_up_replica(self, src_node_idx, added_node_idx, num_active_partitions,
+        server_node_hostname = server_node.account.ssh_hostname
+        return self.get_host(server_node_hostname, waltz_server_cluster.jetty_port)
+
+    def scale_up_server(self, num_active_partitions, txn_per_client, num_clients, interval, timeout, num_added_nodes, delay_before_torture):
+        """
+        A validate function to test scaling up number of servers.
+
+        :param num_active_partitions: Number of active partitions
+        :param txn_per_client: Number of transactions per client
+        :param num_clients: Number of total clients
+        :param interval: Average interval(millisecond) between transactions
+        :param timeout: Test timeout
+        :param num_added_nodes: Number of extra added server nodes
+        :param delay_before_torture: Delay in seconds before new nodes are added after transaction processing starts
+        :returns: Validation result
+        """
+
+        port = self.waltz_storage.port
+        admin_port = self.waltz_storage.admin_port
+        storage_node = self.waltz_storage.nodes[randrange(len(self.waltz_storage.nodes))]
+        storage = self.get_host(storage_node.account.ssh_hostname, admin_port)
+        partition = randrange(num_active_partitions)
+
+        server_cli = ServerCli(self.verifiable_client.nodes[0], self.client_config_path)
+        added_server_nodes = self.get_server_service(int(self.zk_cfg['ClusterNumPartitions']), num_added_nodes)
+
+        original_server = self.get_random_server_node_instance(self.waltz_server)
+        added_server = self.get_random_server_node_instance(added_server_nodes)
+
+        # Step 1: Produce transactions with current cluster.
+        cmd = self.client_cli.validate_txn_cmd(self.log_file_path, num_active_partitions, txn_per_client, num_clients,
+                                               interval)
+        self.verifiable_client.start(cmd)
+        wait_until(lambda: self.is_max_transaction_id_updated(storage, port, partition, -1), timeout_sec=timeout)
+
+        # Step 2: Check partition assignment
+        original_server_num_assigned_partitions_before_test = self.get_number_of_partitions_assigned_to_server(server_cli, original_server)
+        sleep(delay_before_torture)
+
+        # Step 3: Start new server nodes
+        added_server_nodes.start()
+
+        # Step 4: Get partition assignments after adding new server nodes
+        original_server_num_assigned_partitions_after_test = self.get_number_of_partitions_assigned_to_server(server_cli, original_server)
+        added_server_num_assigned_partitions_after_test = self.get_number_of_partitions_assigned_to_server(server_cli, added_server)
+
+        assert added_server_num_assigned_partitions_after_test > 0 and \
+               (original_server_num_assigned_partitions_after_test < original_server_num_assigned_partitions_before_test or
+                original_server_num_assigned_partitions_before_test == 1), \
```

Comment:
Why `original_server_num_assigned_partitions_before_test == 1` in the assert?

Reply:
let's imagine a scenario where we have 4 partitions and 3 server nodes. 2 nodes will have only one partition assigned and this number won't decrease by adding an extra server node.